ion-schema/src/model/constraints/all_of.rs

```diff
@@ -0,0 +1,173 @@
+    use crate::{ISL_1_0, ISL_2_0};
+
+    use super::*;
+
+    #[test]
+    fn test_builder() {
+        let type_ = TypeDefinitionBuilder::<ISL_1_0>::new()
+            .all_of((
+                // This is a heterogeneous "list" (really a tuple) of things that can become a type arg.
+                "foo_type",
+                TypeDefinitionBuilder::new().build(),
+                TypeReference::imported("bar.isl", "baz"),
+                "quux_type".or_null(true),
+            ))
+            .build();
+
+        assert_eq!(
+            type_.constraints().cloned().collect::<Vec<_>>(),
+            vec![AllOf::new(vec![
+                TypeArgument::from_type_ref(
+                    NullabilityModifier::Nothing,
+                    TypeReference::from("foo_type")
+                ),
+                TypeArgument::from_type_def(
+                    NullabilityModifier::Nothing,
+                    TypeDefinition::new(vec![], vec![])
+                ),
+                TypeArgument::from_type_ref(
+                    NullabilityModifier::Nothing,
+                    TypeReference::imported("bar.isl", "baz")
+                ),
+                TypeArgument::from_type_ref(
+                    NullabilityModifier::Nullable,
+                    TypeReference::from("quux_type")
+                ),
+            ],)
+            .into()]
+        );
+
+        let type_ = TypeDefinitionBuilder::<ISL_2_0>::new()
+            .all_of(("foo_type", "quux_type".or_null(true)))
+            .build();
+
+        assert_eq!(
+            type_.constraints().cloned().collect::<Vec<_>>(),
+            vec![AllOf::new(vec![
+                TypeArgument::from_type_ref(
+                    NullabilityModifier::Nothing,
+                    TypeReference::from("foo_type")
+                ),
+                TypeArgument::from_type_ref(
+                    NullabilityModifier::NullOr,
+                    TypeReference::from("quux_type")
+                ),
+            ],)
+            .into()]
+        );
+    }
+
+    #[test]
+    fn test_equality() {
+        assert_eq!(
+            TypeDefinitionBuilder::<ISL_1_0>::new()
+                .all_of(("a", "b", "c"))
+                .build(),
+            TypeDefinitionBuilder::<ISL_1_0>::new()
+                .all_of(("b", "c", "a"))
+                .build(),
+        );
+        assert_eq!(
+            TypeDefinitionBuilder::<ISL_1_0>::new()
+                .all_of(("a", "b", "c", "a", "a"))
+                .build(),
+            TypeDefinitionBuilder::<ISL_1_0>::new()
+                .all_of(("a", "b", "a", "c", "a"))
+                .build(),
+        );
+        assert_ne!(
+            TypeDefinitionBuilder::<ISL_1_0>::new()
+                .all_of(("a", "b", "c", "a"))
+                .build(),
+            TypeDefinitionBuilder::<ISL_1_0>::new()
+                .all_of(("a", "b", "c", "b"))
+                .build(),
+        );
```

Comment:
Maybe add a comment highlighting that this is checking for Bag (not Set) equivalence?